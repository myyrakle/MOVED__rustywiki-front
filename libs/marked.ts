import unified from 'unified';
import markdown from 'remark-parse';
import html from 'remark-html';
import * as remarkWikiLink from 'remark-wiki-link';
import * as footnotes from 'remark-footnotes';
const md = unified()
  .use(markdown)
  .use(footnotes.default)
  .use(remarkWikiLink.wikiLinkPlugin, {
    hrefTemplate: (permalink: string) => `/d/wiki/${permalink}`,
    aliasDivider: '|',
  })
  .use(html, { sanitize: true });

export const customMarked = md;
