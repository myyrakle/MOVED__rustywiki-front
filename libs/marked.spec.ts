import { customMarked } from './marked';

describe('markdown render to html', () => {
  describe('위키링크', () => {
    it('[[contents]] 렌더링이 a element로 되어야 한다.', async () => {
      const res = await customMarked.process('[[테스트]]');
      expect(String(res)).toMatchSnapshot();
    });

    it('[[content|alias]] alias가 대신 나와야 한다.', async () => {
      const res = await customMarked.process('[[content|alias]]');
      expect(String(res)).toMatchSnapshot();
    });
  });

  describe('footNotes', () => {
    it('[^1] 각주를 렌더링시 [^1] : 각주 링크가 서로 참조해야 한다', async () => {
      const res = await customMarked.process(
        `각주,[^longnote] lorem ipsum\n이것은 각주여[^1]\n[^longnote]: 각주입니다\n[^1]: 각주입니다`
      );
      expect(String(res)).toMatchSnapshot();
    });
  });

  describe('html xss', () => {
    it('허용되지 않은 html에 대해서는 위생화가 되어야 한다.', async () => {
      const res = await customMarked.process(
        `# lorem <div>ipsum</div> \n content`
      );
      expect(String(res)).toMatchSnapshot();
    });
    it('허용되지 않은 script에 대해서는 위생화가 되어야 한다.', async () => {
      const res = await customMarked.process(
        `# lorem <div>ipsum</div> \n content <script>alert("x")</script>`
      );
      expect(String(res)).toMatchSnapshot();
    });
  });
});
