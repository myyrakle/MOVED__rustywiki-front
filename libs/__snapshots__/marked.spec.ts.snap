// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`markdown render to html footNotes [^1] 각주를 렌더링시 [^1] : 각주 링크가 서로 참조해야 한다 1`] = `
"<p>각주,<sup id=\\"user-content-fnref-longnote\\"><a href=\\"#fn-longnote\\">longnote</a></sup> lorem ipsum
이것은 각주여<sup id=\\"user-content-fnref-1\\"><a href=\\"#fn-1\\">1</a></sup></p>
<div>
<hr>
<ol>
<li id=\\"user-content-fn-longnote\\">각주입니다<a href=\\"#fnref-longnote\\">↩</a></li>
<li id=\\"user-content-fn-1\\">각주입니다<a href=\\"#fnref-1\\">↩</a></li>
</ol>
</div>
"
`;

exports[`markdown render to html html xss 허용되지 않은 html에 대해서는 위생화가 되어야 한다. 1`] = `
"<h1>lorem ipsum</h1>
<p>content</p>
"
`;

exports[`markdown render to html html xss 허용되지 않은 script에 대해서는 위생화가 되어야 한다. 1`] = `
"<h1>lorem ipsum</h1>
<p>content alert(\\"x\\")</p>
"
`;

exports[`markdown render to html 위키링크 [[content|alias]] alias가 대신 나와야 한다. 1`] = `
"<p><a href=\\"/d/wiki/content\\">alias</a></p>
"
`;

exports[`markdown render to html 위키링크 [[contents]] 렌더링이 a element로 되어야 한다. 1`] = `
"<p><a href=\\"/d/wiki/테스트\\">테스트</a></p>
"
`;
